### PR TITLE
[Feature:InstructorUI] Update Course ID from Manage Sections

### LINKED ISSUE #8770: Add "course_id" to the database
**What problem are you trying to solve with Submitty**
Each course section has an ID number with the registrar.  At RPI it is a 5 digit number that is unique for the semester and we call it a CRN.  We would like to store this information in the database and have it available in Rainbow Grades.

**Describe the way you'd like to solve this problem**

Step 1:

In the main Submitty database, let's add a column named 'course_id' to the courses_registration_sections table.
And in the course database, let's add a column named 'course_id' to the sections_registration table.

This is an optional piece of information, and different schools might have different types (numbers vs strings) or other requirements (uniqueness, etc.). So at this time, let's not attempt to put these requirements on the database.  Let's set the type of this field as 'character varying' (string).  It can have any length (# of characters) and is allowed to be empty string.

This field should be automatically synced between the Submitty database & each course database via database triggers.

Step 2:

When student data is exported for Rainbow Grades, in addition to the field, e.g., "registration_section" : "5", let's add "course_id" : "12345".

Step 3:

Let's display the section ID on the Manage Sections page for the course in the top section under "Manage Registration Sections", add a column to the "Current Registration Section" table.

Step 4:

It would be nice to add a feature to set the course id for a section from the Manage Sections page.  This information should update both the main database and the course database via triggers from step 1.  I expect we at RPI will mostly be inputting the section data via sysadmin scripts with data from the registrar.

### PR DESCRIPTION
**Pull request by Rkoester47**
### Why is this Change Important & Necessary?
Fixes #8770 
Currently, there is no way to update the Course ID for a given section on the Manage Sections page. 

### What is the New Behavior?
Now, there are pencil buttons next to the Course ID for each section. These buttons will open a popup so that an instructor can change the Course ID from this page. This change will update the Course ID in both the Submitty Database and Course Database.

The styling for the pencil buttons was taken directly from the User Profile page pencil buttons.

### What steps should a reviewer take to reproduce or test the bug or new feature?
1. Login as Instructor
2. Go to Sample > Manage Sections
3. Click the pencil button next to a course section
4. Update the Course ID with a new number and save - ensure the success banner displays.
5. Attempt to update a section to have the same Course ID as another section - error banner should display.

### Other information
This is not a breaking change.

### CODE REVIEW

**Review by aconfo — CHANGES_REQUESTED** [review]
I primarily conducted a functionality test. When testing the changed work, immediately I noticed that the "pencil" icons to change the course ID were present. Clicking on it opened up a small window that allows course ID to be changed, at which point I conducted a variety of tests for edge cases. It successfully limited me to inputting only a 5 digit number, with no letters working which is all good. It also prevented me from inputting an empty ID. There was however, one problem I noticed when testing. If, let's say, I make a course ID 00002. Now when I try to add a new course ID to another course it would deny me from adding 00002 as its ID, which is good and intended. The problem arose after I removed 00002 as an ID from the course I assigned it to, where after removing it, I still couldn't add it as a valid ID as seen below: 

Image 

Image 

Image
This could pose a significant problem and prevent course ID's from being edited, and I would recommend this be changed to only restrict course ID's being used.

File: site/app/controllers/admin/UsersController.php [hunk 1]
```diff
@@ -606,6 +606,42 @@ public function updateRegistrationSections() {
         $this->core->redirect($return_url);
     }
 
+    #[Route("/courses/{_semester}/{_course}/sections/update_course_id", methods: ["POST"])]
+    public function updateCourseId(): JsonResponse {
+        try {
```
Comment (JManion32) [inline site/app/controllers/admin/UsersController.php]: You shouldn't have to verify the CSRF token here, I believe all POST requests are processed in `WebRouter.php` before being directed to controller endpoints.

File: site/app/controllers/admin/UsersController.php [hunk 2]
```diff
@@ -606,6 +606,42 @@ public function updateRegistrationSections() {
...
+
+        if (!preg_match('/^\d{5}$/', $course_id)) {
+            return JsonResponse::getErrorResponse('Course ID must be a 5-digit number');
+        }
+
+        if ($this->core->getQueries()->courseIdExists($course_id, $section_id) > 0) {
+            return JsonResponse::getErrorResponse('That Course ID is already in use.');
+        }
+
+        $semester = $this->core->getConfig()->getTerm();
```
Comment (JManion32) [inline site/app/controllers/admin/UsersController.php]: Always use `term` instead of `semester`, since we want to keep our site (and codebase) as general as possible.

File: site/public/css/rotatingsectionsform.css [hunk 3]
```diff
@@ -34,3 +34,25 @@ table caption {
 table caption .subcaption {
     font-size: 1rem;
 }
+
+.edit-course-btn {
```
Comment (JManion32) [inline site/public/css/rotatingsectionsform.css]: I understand these are the same classes from `user-profile.css`, so why don't we abstract them into a file that both the rotating sections form and the user profile settings page can pull from? I think the best option is `server.css`. This will keep keep instances of `!important` in our codebase to a minimum, and allow for reuse of the styles for a common button. I can't imagine an instance where we would want to change the style of 1 and not the other.

File: site/app/templates/admin/users/RotatingSectionsForm.twig [hunk 4]
```diff
@@ -263,4 +287,83 @@
             </div>
         </div>
     </form>
+    <script>
```
Comment (JManion32) [inline site/app/templates/admin/users/RotatingSectionsForm.twig:290]: This is a lot of new inline JavaScript. I think we should abstract all of it into a `rotating-sections.js` file if possible.

File: site/app/libraries/database/DatabaseQueries.php [hunk 5]
```diff
@@ -3210,6 +3210,29 @@ public function deleteRegistrationSection($section) {
         return $this->submitty_db->getRowCount();
     }
 
+    public function updateCourseSectionId(string $section_id, string $course_id): void {
+        $this->submitty_db->query("UPDATE courses_registration_sections SET course_section_id = ? WHERE registration_section_id = ?", [$course_id, $section_id]);
+
+        $semester = $this->core->getConfig()->getTerm();
```
Comment (JManion32) [inline site/app/libraries/database/DatabaseQueries.php]: Use `term` instead of `semester`

**Comment by dagemcn** [comment]
I did a functionality review on this, the previous issue that @aconfo mentioned seems to be resolved, though I can't speak to the changes requested for the code. This is a great quality of life improvement for instructors. I tested quite a few special characters along with alt-code characters. Everything is appears to be working as intended. One thing I noticed is a redundant close button, although I'm not sure if that is the correct style for a pop up menu like this. 
image
Someone more knowledgeable may be able to comment on this.

**Comment by JManion32** [comment]
> I did a functionality review on this, the previous issue that @aconfo mentioned seems to be resolved, though I can't speak to the changes requested for the code. This is a great quality of life improvement for instructors. I tested quite a few special characters along with alt-code characters. Everything is appears to be working as intended. One thing I noticed is a redundant close button, although I'm not sure if that is the correct style for a pop up menu like this. image Someone more knowledgeable may be able to comment on this.

@dagemcn We use a template for popups so we don't have to build a new one from scratch every time we need one. The `close` button in the top right is always there, and I believe the one on the bottom can be customized, but usually just stays as the default.

**Comment by Rkoester47** [comment]
> I did a code review. I left some comments throughout the code. Great job with this so far!

Thank you for the feedback. I implemented the requested changes, and also please take a look at my reply to you on [PR#12412](https://github.com/Submitty/Submitty/pull/12412), about merging my two PRs before moving the Javascript to its own .js file. I propose moving the add section input to a better spot in #12429 , and  then a refactor PR for the page in the future.

File: site/public/css/server.css [hunk 6]
```diff
@@ -2212,3 +2212,26 @@ End of slider switch
     background-color: var(--standard-light-red);
     color: var(--default-black);
 }
+
+.icon {
+    /* moved from user-profile.css */
```
Comment (JManion32) [inline site/public/css/server.css]: Remove this comment
